### DISCUSSION TITLE
How to debug not executed(?) pre-mount hook

### DISCUSSION THREAD

**Discussion by ceisserer** [dracutdevs/dracut#2061] [General]
I would like to register a pre-mount hook in order to perform a few changes on rootfs right before it gets mounted for real.

What I did was:
- create a folder /usr/lib/dracut/modules.d/40extsync with module-setup.sh and extsync.sh and set permissions to a+x
- module-setup.sh contains check and depends functions both returning 0, as well as install with: inst_hook pre-mount 91 "$moddir/extsync.sh"
- extsync.sh for now only mounts the root-fs manually, creates a file and umounts it (just to test whether it is executed):
- ran dracut -f

However after booting, rootfs does not contain the file the script should be been created, even through the script script is included in the initrd image.What would be the next steps to find out what is going wrong? I already added "rd.debug log_buf_len=1M" to the kernel line, however I don't see any messages about something going wrong. What would be the next logical issues to look for?

Thanks, Clemens

extsync.sh for now performs the following steps:

#!/bin/sh

/usr/bin/mkdir mp;
/usr/bin/mount /dev/sda3 mp;
/usr/bin/echo "it works" > mp/yes.txt;
/usr/bin/umount mp;
/usr/bin/rm -r mp;

**Comment by LaszloGombos** [comment]
@ceisserer 

>"rd.debug" ..I don't see any messages about something going wrong

Would you be able to post the log here ? In the log you should see your script called.

Are all the dependencies of the "extsync.sh" script itself are also included in initramfs ? 

You can also try rd.break=mount

Ps.: You do not need to add a new module just to register a hook. You can just add the hook script in dracut command line

**Reply by ceisserer** [reply]
Hello Laszlo,

thanks for your suggestions, there indeed was some output, maybe I looked at the wrong place.

The problem why the script did not work, was that by the time the pre-mount hook was called, the device was not ready. For now I was able to solve it via an ugly hack - a bash loop waiting for the device file to appear.

However, my real goal still seems to be far away:
I would like to boot from USB and in case I am on my own PC, sync all files from USB to a NVME partition and instead use this one as rootfs. The sync operation is done via btrfs snapshots and send/receive and works extremly well.
Otherwise the system should continue booting from USB.

However, it seems I can't tell dracut to actually re-mount the nvme parition instead of what has been specified as rootfs= kernel parameter.

kernel cmd line is: rootfs=/dev/sda3 (the usb device)
1. boot with rootfs=/dev/sda3

in pre-mount hook (so / has not been mounted)
2. sync sda3 to /dev/nvme0n1p5 
3. rewrite fstab on both /dev/sda3 and /dev/nvme0n1p51 to mount /dev/nvme0n1p5 as /

However after booting, the sync-operation has been successfully completed and fstab is successfully rewritten on both devices, the system is still running with /dev/sda3 as rootfs.

So I wonder, is there some way to tell dracut via some hook to not continue with the specified rootfs but instead use another device.
I had a look at the modules, but the seem to be (of course) written in a rather generic way, making it quite hard for me to grasp :/

**Comment by LaszloGombos** [comment]
> I would like to boot from USB and in case I am on my own PC, sync all files from USB to a NVME partition 

What suppose to happen if you shut down the computer ? Sync back ? 

Can you try to solve this at the storage level ? Maybe btrfs raid or bcache ? 
If you do this than there is only ever one root (and several drives as backing store for that one root).
